### DISCUSSION TITLE
API to control wlr-screencopy at runtime.

### DISCUSSION THREAD

**Discussion by Fox2Code** [CuarzoSoftware/Louvre#53] [Ideas]
The idea I have is simple, allow compositor to implement handlers for wlr-screencopy to allow compositors to implement a permission system, an example of possible implementation for a compositor could be to only send a pitch black buffer unless permission to capture screen is granted via a dialog box.

Well having something for that in the documentation at least would be cool.

**Comment by Fox2Code** [comment]
Another idea of possible implementation is showing which wayland process is capturing the Window with a visual indicator.

**Comment by ehopperdietzel** [comment]
Yes, the API is available, and I'll elaborate on it in a new tutorial chapter. It's as straightforward as this:

```cpp
void Output::paintGL()
{
    for (LScreenshotRequest *req : screenshotRequests())
        req->accept(true); // Use false or ignore to deny
}
```

Additionally, each `LScreenshotRequest` provides the `LClient` and the desired output rectangle to be captured. This information can be used for client authentication. However, I believe the optimal approach is to use the wlr desktop portal, allowing only pipewire to bind to the `wlr-screencpy-manager` global. Although I'm not entirely certain, I believe the portal handles displaying a dialog for accepting or denying requests. Regardless, by default, all requests are ignored. In `louvre-views` and `louvre-weston-clone`, I simply grant permission to all, as they serve as mere examples.

**Reply by Fox2Code** [reply]
The best is for Louvre to implement everything needed and for compositor developers to forget about that, the reason I support Louvre is because I believe in what it try to achieve, parallelism and easy to use APIs. (Meaning I believe Louvre should probably handle portals too when relevant, because xdg dbus portals are a part of wayland now)

Also I think requests should be accepted by default by the compositor automatically if there is no override.

**Reply by ehopperdietzel** [reply]
> The best is for Louvre to implement everything needed and for compositor developers to forget about that.

Hmm, Louvre is handling all copies, cursor compositing, damage tracking, and synchronization. The user only needs to decide whether to allow a request or not. I don't think it can get any simpler than that without losing flexibility.

> Also I think requests should be accepted by default by the compositor automatically if there is no override.

I think that would be unsafe, someone not familiar with this feature might leave it enabled without realizing it. Additionally, enabling it is just as simple as adding two lines of code. `LOutput::paintGL()` will probably always be overridden, as it is where rendering is performed for a display.

**Reply by ehopperdietzel** [reply]
Also, the requests are for a single frame only, so if one is denied, it doesn't mean the "recording session" is canceled.

**Reply by Fox2Code** [reply]
My idea was sending a black frame possibly only exposing the captured buffer size when capture is denied.
For the old protocol, compositors could use a 1s cooldown to keep software as "capturing screen" so it's not a big issue either.

Also you don't have to rush things, as long as XWayland isn't a thing, I'll say Louvre is still incomplete, and I think you will be able to make massive breaking changes without upsetting peoples too much.

**Reply by ehopperdietzel** [reply]
> My idea was to send a black frame, possibly only exposing the captured buffer size when capture is denied.

Why? So that the session is not cancelled?

> For the old protocol, compositors could use a 1s cooldown to keep software "capturing screen," so it's not a big issue either.

Good idea.

> Also, you don't have to rush things. As long as XWayland isn't a thing, I'll say Louvre is still incomplete, and I think you will be able to make massive breaking changes without upsetting people too much.

Yeah, you're right. I think I'll add as many protocols as possible to 2.0.0 before releasing, so it takes a long time to reach 3.0.0.

**Reply by Fox2Code** [reply]
> Why? So that the session is not cancelled?

Yeah, in case the capturing cannot be canceled, or in case the permission dialog box hasn't been selected yet, if the protocol support it, it would be better to notify the application that the user hasn't granted screen capture capabilities to the client yet, but that screen capture may be granted by the user.

**Reply by ehopperdietzel** [reply]
I could implement that, but it would waste system resources since clients would copy and store frames anyway. The ideal scenario would be to allow a single client, or none at all. This can be achieved directly within the `paintGL()` call or by exposing the protocol global to a specific client (which I will explain in the tutorial). Allowing multiple clients to copy the entire screen significantly reduces FPS. For example, with two clients, I achieve an average FPS of 30 (with DMA), with SHM is even worse. 

Therefore, the idea is to allow a single client (e.g. a portal) responsible for providing an authorization modal or similar interface, and then distributing the captured frames among all other permitted applications. 

I am also considering creating a protocol to directly share the screen buffers through DMA, which would involve zero copying from the compositor and result in significantly faster performance.

**Reply by Fox2Code** [reply]
@ehopperdietzel then leverage multithreading for non DMA multiclient copying, like Louvre copy it own buffer into a "staging buffer", which is a buffer that is a copy of the entire screen buffer, with locks and safe multi-threading it's definitely possible.

```
Louvre Rendering
      \/
Frame copy ------------------------\
      \/                          \/
Louvre Rendering | Send Frame to clients
      \/
Frame copy ------------------------\
      \/                          \/
Louvre Rendering | Send Frame to clients
```

This is what I have in mind, but there could still be better ideas.

Pipewire buffers can support DMA-BUF: <https://docs.pipewire.org/page_dma_buf.html>

So it's theoretically possible to make a DMA-BUF screen capture by implementing the `ScreenCast` portal:  
<https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.ScreenCast.html>

Note: Louvre should add a "Supported portals" category in `README.md` at some point.

Note²: Don't rush things, all of this can take it's time, and doesn't have to be out by `2.0.0`

**Reply by ehopperdietzel** [reply]
> @ehopperdietzel then leverage multithreading for non DMA multiclient copying, like Louvre copy it own buffer into a "staging buffer", which is a buffer that is a copy of the entire screen buffer, with locks and safe multi-threading it's definitely possible.

I think I understand your point, but wouldn't that involve more copies than the current implementation? If I understand correctly, what you're suggesting would look like this:

1. Render.
2. Copy from GPU to main memory buffer. Copy count: 1.
3. Copy from main memory buffer to client 1 SHM buffer. Copy count: 2.
4. Copy from main memory buffer to client N SHM buffer. Copy count: N + 1.

The current approach would only make N copies from the GPU to the SHM buffers, which, depending on the GPU or if there is caching involved, could be slower or faster than your suggestion. That would need to be tested. Additionally, some clients want the cursor to be composited, while others don't, adding extra complexity. So basically, the best approach, in my opinion, would be to allow a single client (single GPU -> SHM buffer copy) and let that client be in charge of multiplexing the buffer with other apps, similar to how the portal handles it, I assume, with pipewire.

Now, the idea of the protocol I mentioned earlier would work like this:

Louvre currently, by default, creates 2 framebuffers for each output (or 1 or 3, depending on environments like SRM_RENDER_MODE_ITSELF_FB_COUNT). Those buffers are usually DMA exportable. Also, almost all drivers support hardware cursor planes, so there is no need to render the cursor into the framebuffers. So, with that said, the protocol would function as follows:

1. A single client binds to the global.
3. Louvre sends all the DMA buffers parameters to the client, each with a specific index [0, 1, 2] (assuming triple buffering).
4. Similarly to step 2, send small buffers with the cursor used in a frame.
5. Output render loop:
   - 4.1 Render to fb 0.
   - 4.2 Wait for finish, then display it on screen and bind fb 1.
   - 4.3 Tell the client to copy from fb 0 (and use cursor 0).
   - 4.4 Client copies, performs the cursor composition itself and multiplex with other apps.
   - 4.5 Repeat with fb 1,  fb 2, then fb 0 again and so on...

This would involve zero copying from the compositor, and if using triple buffering, the client would have plenty of time to copy, or perhaps a synchronization primitive could be used. The issue arises if DMA buffers cannot be shared as read-only, in such a case, the client should be a well-trusted client. Additionally, if DMA is not supported, the approach would be similar, but Louvre would share 3 SHM buffers (and would need to perform a GPU > SHM copy each frame).

The second issue arises if the driver does not support hardware cursor planes; the cursor would always be rendered and would appear in captures even if the client doesn't request it. A solution for that could be using 3 extra small buffers to copy the region of the cursor before it is rendered so that the client can clean it.

> Pipewire buffers can support DMA-BUF: https://docs.pipewire.org/page_dma_buf.html

> So it's theoretically possible to make a DMA-BUF screen capture by implementing the ScreenCast portal:
https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.ScreenCast.html

I'm going to look into that, I don't completely understand yet how portals are implemented.

> Note: Louvre should add a "Supported portals" category in README.md at some point.

Ok, sounds good to me.

**Reply by Fox2Code** [reply]
I completely forgot about cursor capture, yeah maintaining many different buffers just for a cursor is meh, even if it can be handled smartly, I am not sure putting over-complicated code on Louvre is a good idea, as feature creep like this can easily make the project unmaintainable.

Implementing portal should be similar to implementing wayland protocols, but with dbus.

Also Pipewire has a permission API system too: <https://docs.pipewire.org/page_access.html>

**Reply by ehopperdietzel** [reply]
> I am not sure putting over-complicated code on Louvre is a good idea, as feature creep like this can easily make the project unmaintainable.

Yeah, it was just an idea, but it actually would be simpler for the compositor and harder to implement for clients.

> Implementing portal should be similar to implementing wayland protocols, but with dbus.

Just tested the wlroots portal and it works 😄

![Louvre_Screenshoot_2024-04-16 19:44:40](https://github.com/CuarzoSoftware/Louvre/assets/14140144/90337c8f-222a-4741-a0e0-93f9e9c13170)

It has a config file for selecting the output chooser (which can be a wayland app/modal) and I think that if it returns an error code the capture is cancelled, so its perfect!

https://man.archlinux.org/man/xdg-desktop-portal-wlr.5

Edit: It actually says:

> To signal that the user has declined screencast, the chooser should exit without anything on stdout.

**Reply by Fox2Code** [reply]
New `xdg-desktop-portal-lvr` using `ext-screencopy-v1` would be a great idea imo, also an API for compositor to define portals would be great too, but I'm a bit worried about compositor security in some way, and the compositor parsing all client inputs, I think as long as a client input is processed off thread, and that it support invalid request data handling, it should be fine.

But remember to take your time, `ext-screencopy-v1` won't be available for `2.0.0` release, better to keep that for `3.0.0` or later

**Reply by ehopperdietzel** [reply]
> New xdg-desktop-portal-lvr using ext-screencopy-v1 would be a great idea imo.

The wlr portal works fine right now, and will probably support `ext-screencopy-v1` in the future (maybe it does already), so creating another one just for Louvre compositors is not required yet, unless there is special functionality we want to add. For example, adding support for the custom screen capture protocol I mentioned. On the other hand, everybody is free to create their own portal if they wish.

> Also, an API for the compositor to define portals would be great too, but I'm a bit worried about compositor security in some way, and the compositor parsing all client inputs. I think as long as a client input is processed off-thread, and it supports invalid request data handling, it should be fine.

There is `wl_display_set_global_filter` to allow/deny clients binding to globals, so for example, `xdg-desktop-portal-wlr` can be started by Louvre using `LLauncher` (to get its PID), and then in the filter callback, we can check if it matches the requesting `LClient` PID and only allow that client to bind to the global. Doing so means that there will only be frame capture requests from the portal and no other client (safe). Also, the output chooser (e.g., an auth modal) is launched by the portal, not by Louvre. It doesn't block Louvre's execution, only if accepted, Louvre will receive frame capture requests from the portal. Handling a frame capture request within `paintGL()` is also non-blocking, it only sets a boolean variable indicating if the frame was accepted, and the copy is done later.

So I think the current API is flexible enough to allow devs to do whatever they want without too much effort. What do you think?

**Reply by Fox2Code** [reply]
Yeah, devs can always copy paste code from the examples, as long as it's abstracted in a way whatever how the client request, the compositor has to handle the permission one time, and with the same piece of code if possible.

Handling of stuff in the rendering loop or `paintGL()` should be avoided, unless it's really simple and unlikely to cause slowdowns, the magic of Louvre is it's multi-threading, recent CPUs just have so many cores.

**Comment by ehopperdietzel** [comment]
There is [this protocol](https://gitlab.freedesktop.org/andri/wayland-protocols/-/blob/zext-screencopy-unstable-v1/staging/ext-screencopy/ext-screencopy-v1.xml) currently being developed, which addresses many issues present in `wlr-screencopy-manager`. For example it introduces the concept of a "session," whereas the current protocol only allows clients to request capturing individual frames without any means to link them together. Once this new protocol is considered stable, I plan to replace the existing one. This will allow for providing an API to allow/deny a session instead of individual frames, which is more suitable.

**Reply by Fox2Code** [reply]
Well, the most important is still that Louvre API is easy to use, the idea is if someone makes something for one version of Louvre, it still work though the next version.
